packages/payment-processor/src/payment/encoder-approval.ts

```diff
@@ -0,0 +1,145 @@
+import {
+  hasErc20ApprovalForProxyConversion,
+  prepareApproveErc20ForProxyConversion,
+} from './conversion-erc20';
+import { hasApprovalErc20ForSwapToPay, prepareApprovalErc20ForSwapToPay } from './swap-erc20';
+import {
+  hasErc20ApprovalForSwapWithConversion,
+  prepareApprovalErc20ForSwapWithConversionToPay,
+} from './swap-conversion-erc20';
+import { getPaymentNetworkExtension } from './utils';
+
+export function encodeRequestApprovalIfNeeded(
+  request: ClientTypes.IRequestData,
+  provider: providers.Provider,
+  from: string,
+  options?: IRequestPaymentOptions,
+): Promise<IPreparedTransaction | void> {
+  if (options && options.swap) {
+    return encodeRequestApprovalWithSwapIfNeeded(request, provider, from, options);
+  } else {
+    return encodeRequestApprovalWithoutSwapIfNeeded(request, provider, from, options);
+  }
+}
+
+export async function encodeRequestApprovalWithoutSwapIfNeeded(
+  request: ClientTypes.IRequestData,
+  provider: providers.Provider,
+  from: string,
+  options?: IRequestPaymentOptions,
+): Promise<IPreparedTransaction | void> {
+  const paymentNetwork = getPaymentNetworkExtension(request)?.id;
+
+  switch (paymentNetwork) {
+    case ExtensionTypes.ID.PAYMENT_NETWORK_ERC20_PROXY_CONTRACT:
+      if (!(await hasErc20Approval(request, from))) {
+        return prepareApproveErc20(request, provider);
+      }
+      break;
+    case ExtensionTypes.ID.PAYMENT_NETWORK_ERC20_FEE_PROXY_CONTRACT:
+      if (!(await hasErc20Approval(request, from))) {
+        return prepareApproveErc20(request, provider);
+      }
+
+      break;
+    case ExtensionTypes.ID.PAYMENT_NETWORK_ANY_TO_ERC20_PROXY: {
+      if (
+        !options ||
+        !options.conversion ||
+        !options.conversion.currency ||
+        options.conversion.currency.type !== RequestLogicTypes.CURRENCY.ERC20
+      ) {
+        throw new Error('Conversion settings missing');
+      }
+      const amount = options.conversion.maxToSpend
+        ? options.conversion.maxToSpend
+        : BigNumber.from(0);
+      if (
+        !(await hasErc20ApprovalForProxyConversion(
+          request,
+          from,
+          options.conversion.currency.value,
+          provider,
+          amount,
+        ))
+      ) {
+        return prepareApproveErc20ForProxyConversion(
+          request,
+          options.conversion.currency.value,
+          provider,
+        );
+      }
+      break;
+    }
+  }
+}
+
+export async function encodeRequestApprovalWithSwapIfNeeded(
+  request: ClientTypes.IRequestData,
+  provider: providers.Provider,
+  from: string,
+  options: IRequestPaymentOptions,
+): Promise<IPreparedTransaction | void> {
+  const paymentNetwork = getPaymentNetworkExtension(request)?.id;
+
+  switch (paymentNetwork) {
+    case ExtensionTypes.ID.PAYMENT_NETWORK_ERC20_FEE_PROXY_CONTRACT:
+      if (options && options.swap) {
+        if (
+          !(await hasApprovalErc20ForSwapToPay(
```

Comment:
I think it's better to keep the check for approval outside of this. Actually we need a generic top level method to approve if needed. But the encoding should not be concerned by it. It should remain a simple, synchronous, method

Reply:
Now there are the following methods (global, with and without swa):
- Approval encoding
- Check for knowing if approval is needed
- Approval encoding if needed